services/api/src/api/routes/rows.py

```diff
@@ -0,0 +1,89 @@
+from starlette.requests import Request
+from starlette.responses import Response
+
+from api.authentication import auth_check
+from api.utils import (
+    Endpoint,
+    MissingRequiredParameterError,
+    ResponseNotFoundError,
+    ApiCustomError,
+    are_valid_parameters,
+    get_json_api_error_response,
+    UnexpectedError,
+    get_json_ok_response,
+    get_json_error_response,
+)
+
+
+import hffs
+
+# https://github.com/apache/arrow/issues/32609
+import pyarrow.parquet as pq  # type: ignore
+import requests
+
+
+def create_rows_endpoint(
+    parquet_processing_step: ProcessingStep,
+    external_auth_url: Optional[str] = None,
+    max_age_long: int = 0,
+    max_age_short: int = 0,
+) -> Endpoint:
+    async def rows_endpoint(request: Request) -> Response:
+        try:
+            dataset = request.query_params.get("dataset")
+            config = request.query_params.get("config")
+            split = request.query_params.get("split")
+            if not are_valid_parameters([dataset, config, split]):
+                raise MissingRequiredParameterError("Parameter 'dataset', 'config' and 'split' are required")
+            MAX_ROWS = 100
+            fromRow = int(request.query_params.get("from", 0))
+            _toRow = request.query_params.get("to")
+            if _toRow is None:
+                toRow = fromRow + MAX_ROWS
+            else:
+                toRow = max(min(int(_toRow), fromRow + MAX_ROWS), fromRow)
+            logging.info("/rows, dataset={dataset}, config={config}, split={split}")
+
+            # if auth_check fails, it will raise an exception that will be caught below
+            auth_check(dataset, external_auth_url=external_auth_url, request=request)
+            try:
+                # get the list of parquet files
+                result = get_response(kind=parquet_processing_step.cache_kind, dataset=dataset)
+                content = result["content"]
+                http_status = result["http_status"]
+                error_code = result["error_code"]
+                if http_status != HTTPStatus.OK:
+                    # TODO: send a proper error message, saying that the parquet files cannot be created
+                    return get_json_error_response(
+                        content=content, status_code=http_status, max_age=max_age_short, error_code=error_code
+                    )
+                parquet_files_paths = [
+                    f"{config}/{parquet_file['filename']}"
+                    for parquet_file in content["parquet_files"]
+                    if parquet_file["split"] == split and parquet_file["config"] == config
+                ]
+                # get the list of the first 100 rows
+                REVISION = "refs/convert/parquet"
+                fs = hffs.HfFileSystem(dataset, repo_type="dataset", revision=REVISION)
+                dataset = pq.ParquetDataset(parquet_files_paths, filesystem=fs)
+                table = dataset.read()
```

Comment:
This downloads the full dataset and is not efficient.

Reply:
I did this some time ago that can help:
```python
import numpy as np
import pyarrow as pa
import pyarrow.parquet as pq

def index(sources, fs=None):
    all_pf = [pq.ParquetFile(source, filesystem=fs) for source in sources]
    rg_offsets = np.cumsum([pf.metadata.row_group(i).num_rows for pf in all_pf for i in range(pf.num_row_groups)])
    rg_readers = [lambda: pf.read_row_group(i) for pf in all_pf for i in range(pf.num_row_groups)]
    return rg_offsets, rg_readers

def query(fromRow, toRow, rg_offsets, rg_readers):
    start_rg, end_rg = np.searchsorted(rg_offsets, [fromRow, toRow], side="right")
    pa_table = pa.concat_tables([rg_readers[i]() for i in range(start_rg, end_rg + 1)])
    offset = fromRow - rg_offsets[start_rg - 1] if start_rg else fromRow
    pa_table = pa_table.slice(offset, toRow - fromRow)
    rows = pa_table.to_pylist()
    return rows
```

At query time it downloads only the necessary row groups. Each row group contains 1000 elements by default though (defined in `datasets`), which may create some issues for datasets with big rows.

Reply:
Opened https://github.com/huggingface/datasets/pull/5565 to be able to control the row group size.

Right now it's always 1,000 rows - but to optimize random access for datasets with big rows (e.g. with images) we can reduce it

Reply:
Thanks. I started to implement it in #875 (well: to copy your code)